go/vt/sqlparser/token.go

```diff
@@ -43,12 +46,29 @@ type Tokenizer struct {
 	partialDDL    *DDL
 	nesting       int
 	multi         bool
+
+	buf     []byte
+	bufPos  int
+	bufSize int
 }
 
 // NewStringTokenizer creates a new Tokenizer for the
 // sql string.
 func NewStringTokenizer(sql string) *Tokenizer {
-	return &Tokenizer{InStream: strings.NewReader(sql)}
+	buf := []byte(sql)
```

Comment:
I'm wondering if it's worth avoiding this copy. The idea would be to have a string and a []byte buffer, and then `next` would read from only one of them depending on which one is initialized. Of course, this would make us lose the scanString optimization, which will revert to the old way.

If this makes sense, we can also look at moving `Instream, buf`, etc. into a separate type that just implements the `next` and `lastChar` functions, which will be a nice and clean separation from the rest of the tokenizer's functionality.

I haven't thought through the full trade-offs yet. Let me know your thoughts.

Reply:
without the scanString optimization, the Tokenizer got slower, because next() could no longer be inlined. If we moved to an implementation where next has to decide to call a io.Reader.Read(..) , or next() is a virtual function (because we have a StringTokenizer, and a BufferTokenizer) (but honestly I'm not sure how Go does virtual functions), then I also think we'd lose some speed.  So I think what you suggest will degrade performance (compared to where we were before this change), but it would make the code a little cleaner.

Three alternatives,
1) We change the buf to be a string, and on the io.Reader flow take the hit of a copy or
2) We use unsafe to read the []byte directly out of the string. (Which is safe for us, since we don't modify the []byte, and strings are immutable).
3) Change all the callers of NewStringTokenizer to actually pass []byte :) I suspect that'll be a pain, but at some point you must be reading the sql from a socket as a []byte and turning it into a string. Lets avoid that step.

Reply:
None of these options give me a good feeling. I also have a 4th option which is an extension of the original suggestion (which I don't like either);
The proposed 'buf reader' could also pull in the efficient part of the scanString code. But that code would look pretty ugly.

Let's go with what you have now.
Still need to review the rest of the code.